### PR TITLE
refactor: migrate errors server controller to typescript

### PR DESCRIPTION
**Pull request by shuli-ogp**
Part of #144

### CODE REVIEW

File: src/app/controllers/errors.server.controller.ts [hunk 1]
```diff
@@ -1,39 +1,41 @@
...
 }
 
 /**
  * Gets Mongo error object and returns formatted String for frontend
- * @param  {Objrct} err MongoDB error object
+ * @param  {IMongoError} err? MongoDB error object
  * @return {String} message - Error message returned to frontend
  */
-exports.getMongoErrorMessage = function (err) {
+export function getMongoErrorMessage(err?: IMongoError | string): string {
```
Comment (karrui) [inline src/app/controllers/errors.server.controller.ts:38]: This whole function seems like it needs a good refactor, maybe we should try to clean up the function?

I don't think we should create the `IMongoError` interface when we already have the typings of the possible errors:

```ts
import { MongoError } from 'mongodb'
import { Error as MongooseError } from 'mongoose'

...
export const getMongoErrorMessage = (
  err?: MongoError | MongooseError | string,
): string => { ... }
```

Comment (karrui) [reply src/app/controllers/errors.server.controller.ts:38]: Suggested refactor:

```ts
export const getMongoErrorMessage = (
  err?: MongoError | MongooseError | string,
): string => {
  if (!err) {
    return ''
  }

  // Handle base Mongo engine errors.
  if (err instanceof MongoError) {
    switch (err.code) {
      case 10334: // BSONObj size invalid error
        return 'Your form is too large to be supported by the system.'
      default:
        return defaultErrorMessage
    }
  }

  // Handle mongoose errors.
  if (err instanceof MongooseError.ValidationError) {
    // Join all error messages into a single message if available.
    const joinedMessage = Object.values(err.errors)
      .map((err) => err.message)
      .join(', ')

    return joinedMessage ?? err.message ?? defaultErrorMessage
  }

  if (err instanceof MongooseError) {
    return err.message ?? defaultErrorMessage
  }

  return err ?? defaultErrorMessage
}
```

File: src/types/error.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+export interface IMongoError {
```
Comment (karrui) [inline src/types/error.ts:1]: Unneeded, as explained above

File: tests/unit/backend/controllers/errors.server.controller.spec.ts [hunk 3]
```diff
@@ -1,65 +1,62 @@
-describe('Errors Controller', () => {
```
Comment (karrui) [inline tests/unit/backend/controllers/errors.server.controller.spec.ts:1]: Remember to update tests

File: src/app/controllers/errors.server.controller.ts [hunk 4]
```diff
@@ -1,39 +1,41 @@
...
 
 /**
  * Private helper for getMongoErrorMessage to return Mongo error in a String
- * @param  {Object} err - MongoDB error object
+ * @param  {IMongoError} err - MongoDB error object
  * @return {String} errorString - Formatted error string
  */
-const mongoDuplicateKeyError = function (err) {
+
+const mongoDuplicateKeyError = function (err: IMongoError): string {
```
[the remark this answers is not in the document]
Comment (karrui) [reply src/app/controllers/errors.server.controller.ts:18]: We can also keep it in if you want, but I'm not a fan of this weird error message interpolation.

**Review by karrui — CHANGES_REQUESTED** [review]
Some changes requested, with the main one being to change it to a pure utility file. 

Please try to think of ways to improve the code too when you are refactoring instead of plainly converting JS to TS code! The main point we are doing this is to reduce tech debt/legacy code after all :D

**Comment by karrui** [comment]
Any updates on this PR?

**Comment by shuli-ogp** [comment]
Closing first to deprioritise, will pick up again after helium

**Pull request closed without merging**